arangod/Cluster/ClusterFeature.cpp

```diff
@@ -1132,3 +1199,112 @@ std::unordered_set<std::string> ClusterFeature::allDatabases() const {
   }
   return allDBNames;
 }
+
+void ClusterFeature::scheduleConnectivityCheck(std::uint32_t inSeconds) {
+  TRI_ASSERT(_connectivityCheckInterval > 0);
+
+  Scheduler* scheduler = SchedulerFeature::SCHEDULER;
+  if (scheduler == nullptr || inSeconds == 0) {
+    return;
+  }
+
+  auto workItem = arangodb::SchedulerFeature::SCHEDULER->queueDelayed(
+      "connectivity-check", RequestLane::INTERNAL_LOW,
+      std::chrono::seconds(inSeconds), [this](bool canceled) {
+        if (canceled) {
+          return;
+        }
+
+        if (!this->server().isStopping()) {
+          runConnectivityCheck();
+        }
+        if (!this->server().isStopping()) {
+          scheduleConnectivityCheck(
+              _connectivityCheckInterval +
+              RandomGenerator::interval(std::uint32_t(3)));
+        }
+      });
+
+  std::lock_guard<std::mutex> guard(_connectivityCheckMutex);
+  _connectivityCheck = std::move(workItem);
+}
+
+void ClusterFeature::runConnectivityCheck() {
+  TRI_ASSERT(ServerState::instance()->isCoordinator() ||
+             ServerState::instance()->isDBServer());
+
+  TRI_ASSERT(_connectivityCheckFailsCoordinators != nullptr);
+  TRI_ASSERT(_connectivityCheckFailsDBServers != nullptr);
+
+  NetworkFeature const& nf = server().getFeature<NetworkFeature>();
+  network::ConnectionPool* pool = nf.pool();
+  if (!pool) {
+    return;
+  }
+
+  if (_clusterInfo == nullptr) {
+    return;
+  }
+
+  // we want to contact coordinators and DB servers, potentially
+  // including _ourselves_ (we need to be able to send requests
+  // to ourselves)
+  auto servers = _clusterInfo->getCurrentCoordinators();
+  for (auto& it : _clusterInfo->getCurrentDBServers()) {
+    servers.emplace_back(std::move(it));
+  }
+
+  LOG_TOPIC("601e3", DEBUG, Logger::CLUSTER)
+      << "sending connectivity check requests to " << servers.size()
+      << " servers: " << servers;
+
+  // run a basic connectivity check by calling /_api/version
+  static constexpr double timeout = 10.0;
+  network::RequestOptions reqOpts;
+  reqOpts.skipScheduler = true;
+  reqOpts.timeout = network::Timeout(timeout);
+
+  std::vector<futures::Future<network::Response>> futures;
+  futures.reserve(servers.size());
+
+  for (auto const& server : servers) {
+    futures.emplace_back(network::sendRequest(pool, "server:" + server,
```

Comment:
I do not know if `network::sendRequest` can run into problems if it gets a stale connection from the connection pool. This could potentially produce false positives, just because connections in the connection pool are closed without the pool noticing...

Reply:
It probably can. But I would hope that the connection pool will someone figure out if a connection is stable or broken, and then automatically close it at least after a while.
Otherwise all sorts of other requests could fail as well.